### ISSUE TITLE
Cancelling Web flow has Default Error Message instead of Cancel Message

### ISSUE THREAD

**Issue by Entreco** [auth0/Auth0.Android#299]
### Description

> When cancelling the Web Authentication flow, by pressing the Back button, the wrong error **message** is shown. The error _code_ is _correct_

**Expected:** 
AuthenticationException: "The user closed the browser app and the authentication was canceled."

**Actual:** 
AuthenticationException: "An error occurred when trying to authenticate with the server."

Since the code itself IS correct, the issue is not so important. 
It's just inconvenient for people experimenting with Auth0 the first time, and the are skimming to the various error cases they need to handle.

The correct message is set in OAuthManager:
```
boolean resume(AuthorizeResult result) {
    // .. more code
    if (result.isCanceled()) {
            //User cancelled the authentication
            AuthenticationException exception = new AuthenticationException(ERROR_VALUE_AUTHENTICATION_CANCELED, "The user closed the browser app and the authentication was canceled.");
            callback.onFailure(exception);
            return true;
        }
    // more code
}
```

But that seems to get lost somewhere further down the line

### Reproduction

> Launch the Web authentication flow
``` 
        auth.withParameters(params.plus(mapOf("prompt" to "login")))
                .withAudience(data.url)
                .withScope("openid offline_access")
                .withScheme(data.scheme)
                .withResponseType(ResponseType.CODE)
                .withCustomTabsOptions(options)
                .start(activity, auth0Callback)
```

> Press Back button  on device
> onFailure gets called as expected:
```
override fun onFailure(exception: AuthenticationException?) {
            Timber.w("error or cancel:$exception") 
            Timber.w("error or cancel:${exception?.code}")
            Timber.w("error or cancel:${exception?.code == "a0.authentication_canceled"}")
        }
```

> Observe that the code is correct, but we see the default error message. Due to this error always printing the default message, we did NOT know that we could distinguish cancel vs errors in the flow.

### Environment

> Please provide the following:

- **Version of this library used:**: 1.19.0
- **Version of the platform or framework used, if applicable:**: Android, all versions
- **Other relevant versions (language, server software, OS, browser):**: Web flow
- **Other modules/plugins/libraries that might be involved:**: -

**Comment by lbalmaceda** [comment]
I think it might be due to this.. The `AuthenticationException` class extends `Exception` and tries to satisfy its parent class by passing the exception message. 

In this particular case of instantiating it with an error code and description, the message does not take the value of the description but the default value. See here https://github.com/auth0/Auth0.Android/blob/master/auth0/src/main/java/com/auth0/android/authentication/AuthenticationException.java#L55.

This is definitely something that can be improved but I wonder if users are checking the message today as a way to tell which error type this is. Our recommended way nevertheless, is to check the error using the public getters on that class. For a canceled scenario, that would be with this method: https://github.com/auth0/Auth0.Android/blob/master/auth0/src/main/java/com/auth0/android/authentication/AuthenticationException.java#L170

I hope that makes sense.

**Comment by Entreco** [comment]
Makes sense. We are not blocked by this issue at all.
Feel free to close it, or mark it as an improvement, I'll leave that up to you!

**Comment by Canato** [comment]
@poovamraj  @lbalmaceda, we faced a similar situation where we wanted to avoid showing an error message for the user on back navigation.

Is it ok if I open a PR (I can open a new issue, too) with one or the two solutions?

1. Make `ERROR_VALUE_AUTHENTICATION_CANCELED` public so we can check and filter its value instead of having it hardcoded in case the value changes?
1. Or, a more elegant solution, I heritage the `AuthenticationException` class to have `AuthenticationCanceledException`. Looking like it:

```kt 
public class  AuthenticationCanceledException: AuthenticationException

override fun onFailure(error: AuthenticationException) {
    when(error) {
         is AuthenticationCanceledException -> doNothing()
         else -> showErrorMessage(error)
    }
}
```

Any suggestions are welcoming.
